lib/Target/Btor/BtorToBtorIRTranslation.cpp

```diff
@@ -0,0 +1,781 @@
+        }
+        break;
+        case BTOR2_TAG_zero: {
+             auto opType = builder.getIntegerType( line->sort.bitvec.width );
+             res = builder.create<btor::ConstantOp>(unknownLoc, opType, 
+                                                builder.getIntegerAttr(opType, 0));
+        }
+        break;
+        case BTOR2_TAG_bad:
+            res = builder.create<btor::AssertNotOp>(unknownLoc, 
+                                        cache.at(kids[0]));
+        break;
+
+        // indexed ops
+        case BTOR2_TAG_slice: {
+            auto operandWidth = reached_lines.at( kids[0] )->sort.bitvec.width;
+            auto opType = builder.getIntegerType( operandWidth );
+            assert( operandWidth > kids[1] && kids[1] >= kids[2] );
+            auto resType = builder.getIntegerType( kids[1] - kids[2] + 1 );
+
+            auto u = builder.create<btor::ConstantOp>(unknownLoc, opType, 
+                                                builder.getIntegerAttr(opType, kids[1]));
+            auto l = builder.create<btor::ConstantOp>(unknownLoc, opType, 
+                                                builder.getIntegerAttr(opType, kids[2]));
+            res = builder.create<btor::SliceOp>(unknownLoc, resType, cache.at( kids[0]), 
+                                                u->getResult(0), l->getResult(0));
+        }
+        break;
+        case BTOR2_TAG_sext: 
+            res = builder.create<btor::SExtOp>(unknownLoc, 
+                                            cache.at(kids[0]),
+                                            builder.getIntegerType( 
+                                                line->sort.bitvec.width));
+        break;
+        case BTOR2_TAG_uext: 
+            res = builder.create<btor::UExtOp>(unknownLoc, 
+                                            cache.at(kids[0]),
+                                            builder.getIntegerType( 
+                                                line->sort.bitvec.width));
+        break;
+
+        // ternary ops
+        case BTOR2_TAG_ite:
+            res = builder.create<btor::IteOp>(unknownLoc, 
+                                        cache.at(kids[0]),
+                                        cache.at(kids[1]),
+                                        cache.at(kids[2]));
+        break;
+
+        // unmapped ops
+        case BTOR2_TAG_read:
+        case BTOR2_TAG_constraint:
+        case BTOR2_TAG_init: 
+        case BTOR2_TAG_input:
+        case BTOR2_TAG_next:
+        case BTOR2_TAG_sort:
+        case BTOR2_TAG_state:
+        case BTOR2_TAG_fair:
+        case BTOR2_TAG_justice:
+        case BTOR2_TAG_output:
+        case BTOR2_TAG_write:
+        default:
+            break;
+    }
+    return res;
+}
+
+void cleanupCreatedLines() {
+    for( auto it = reached_lines.begin(); it != reached_lines.end(); ++it ) {
+        if( it->first < 0 ) {
+            Btor2Line *line = it->second;
+            delete [] line->args;
```

Comment:
I think this is memory cleanup... If possible, I would avoid `delete` and use `unique_ptr` instead to express ownership

Reply:
The function `createNegateLine` creates a new `Btor2Line` so that we can handle btor2 lines that have the following form: `13 ite 2 -3 10 11`

Storing a `unique_ptr` in the cache results in a type mismatch, hence the explicit memory management. Am I looking at this right?

Reply:
This we'll have to discuss. Alternative is to allow API to take a boolean to indicate that negation needs to be added in mlir, instead of doing on-the-fly transformation and extra memory management.